### ISSUE TITLE
No buttons in WP calc below API 23 (Android 6.0)

### ISSUE THREAD

**Issue by Matze2** [cgeo/cgeo#6704]
##### Detailed steps causing the problem:
- Edit Waypoint Coordinates
- Press "Calculate Coordinates"

##### Actual behavior after performing these steps:
No coordinate buttons except of the (N)orth button and the (E)ast button.
The area is grey.
See attached screenshot. Screenshot with german language but neither doesn't work in English.

##### Expected behavior after performing these steps:
The YouTube video explaining the new coordinate calculator shows button on each coordinate component.

##### Version of c:geo used:
2017.08.22-NB-3917990

##### Is the problem reproducible for you?
Yes

##### System information:
```
--- System information ---
Device: D5503 (D5503, Sony)
Android version: 5.1.1
Android build: 14.6.A.1.236
c:geo version: 2017.08.22-NB-3917990
Google Play services: disabled - 11.3.02 (236-161239932)
Low power mode: inactive
Compass capabilities: yes
Rotation vector sensor: present
Orientation sensor: present
Magnetometer & Accelerometer sensor: present
Direction sensor used: orientation
Hide own/found: true
Map strategy: detailed
HW acceleration: enabled (default state)
System language: de_DE
System date format: dd.MM.yyyy
Debug mode active: no
System internal c:geo dir: /data/data/cgeo.geocaching (2,2 GB free) internal
User storage c:geo dir: /storage/emulated/0/cgeo (2,2 GB free) external non-removable
Geocache data: /storage/sdcard1/Android/data/cgeo.geocaching/files/GeocacheData (7,7 GB free) external removable
Database: /storage/emulated/0/Android/data/cgeo.geocaching/files/databases/data (39,7 MB) on user storage
Fine location permission: granted
Write external storage permission: granted
Geocaching sites enabled:
   geocaching.com: Logged in (Anmeldung OK) / BASIC
Geocaching.com date format: MM/dd/yyyy
Installed c:geo plugins:  none
--- End of system information ---
```

##### Other comments and remarks:
Screenshot of the calculator page attached

**Comment by Matze2** [comment]
Here is the screenshot.

![screenshot_2017-08-22-20-45-18](https://user-images.githubusercontent.com/12542802/29583236-b6c3f07c-877f-11e7-8d08-fffe8c7dfcfb.png)

**Comment by Lineflyer** [comment]
Thanks for your valueable report, have never seen this happening before.

From which cache did you start the calculator?
What coordinates have initially been shown on the windows where you selected "Calculate coordinates"

**Comment by Matze2** [comment]
From a Multi cache in my case GC78Y21.
But it is the same behavior with all caches. 
I do

- Add Waypoint (there it shows my current coordinates)
- Type on coordinates (also here shows my current coordinates)
- Press "Calculate coordinates"

It looks exactly as in the screenshot.

The buttons "N" and "E" switch to "S" and "W" when pressed. That's all what I can do.

Best regards,
Matze

**Comment by Matze2** [comment]
By the way, I am really looking forward to use this feature.

Caching with kids might become stressful since they are impatient when it comes to calculating the coordinates for the next waypoint and under time pressure you tend to make mistakes which is even more frustrating.
With this feature you can prepare all waypoints in advance and have a nice caching experience outside.
I am using nightlies anyway so I can immediately install a new NB whenever you have something to test.

Thanks for your effort,
Matze

**Comment by Lineflyer** [comment]
@slartibartfast Any idea?

**Comment by Matze2** [comment]
Is it maybe related to Android 5.1?
Note again that the North/East buttons are available and working.

**Comment by Lineflyer** [comment]
You could also try the following:
- Switch to white theme in Settings - Appearance and recheck how it looks on white theme
- What happens if you touch the empty fields. Any reaction?
- What happens if you use the coordinate format selection (on topmost) and select another coordinate format?

**Comment by Matze2** [comment]
See screenshots for DDD°MM.MMM' and DDD.DDDDD'.
The buttons are simply not there. Therefore also no reaction when I type anywhere in that area.

![screenshot_2017-08-23-13-47-47](https://user-images.githubusercontent.com/12542802/29615405-22756c22-880e-11e7-8058-21cb96dfdbb1.png)
![screenshot_2017-08-23-13-55-37](https://user-images.githubusercontent.com/12542802/29615411-265fad98-880e-11e7-95f3-bb72128fc2b9.png)

**Comment by Matze2** [comment]
Just to note it.
I use c:geo heavily and never encountered missing buttons or any strange behavior concerning buttons or input fields in any of the dialogs.

**Comment by S-Bartfast** [comment]
That is most peculiar.  I have no idea what could be going wrong for you there.

I just tried GC78Y21 on both my real and emulated devices and it works fine:
![image](https://user-images.githubusercontent.com/6573522/29621048-caf53904-885e-11e7-9c8d-2d28a0565bc8.png)

Have you tried rotating into landscape mode and back to portrait?  I can't really imagine how that would help, but I can't imagine how you arrived at the problem in the first place so anything's worth a try ;)

I take it when you opened the calculator from the 'Input Coordinates' screen you had your current coordinates displayed (which it does by default).  Can I suggest trying to click the "Cache coordinates" button to display the cache's coordinates on the input screen (48° 31.829', E 011° 31.300') and see if that help's the buttons show.  Again, I can't imagine how that would help but we might as well try.

One thing that's curious in your screenshot is the spacing of the format characters.  The layout seems to be making space for buttons that are wider than they should be.

Here is a picture of the two images superimposed on top of each other:
![image](https://user-images.githubusercontent.com/6573522/29622273-1d266a06-8862-11e7-8ab7-b8295f0e4142.png)
It can be seen that the white text symbols are more tightly packed than the darker text symbols.  As to why that has happened is anyone's guess but it definitely hints at something going amiss.

**Comment by Matze2** [comment]
Yes I understood that my device is somehow special. But probably there is a handful people who really tried that feature until now.

I am willing to help to nail down the problem, just tell me what I should try.

To your questions:
- rotating was one of the first things I tried of course. It just made the area wider but no buttons
- using cache coordinates instead of mine also does not make a difference
- about the layout differences: you are right it is more condensed which is even more extreme for DDD°MM'SS.SSS'' ; the last "SSS" area is the widest the first three parts are quite narrow; but the reason could be because it simply occupies less space for "nothing" :)

**Comment by Matze2** [comment]
Ha! Something new which I never tried before:
The "Plain" mode works. If I edit the coordinates and replace a part with an "A" I get the formula editor below.
And first tests seem good.

So the only part that does not work is the **formatted** coordinates editor.

**Comment by S-Bartfast** [comment]
Arrr!
Just set up an emulater running API 22 (Android 5.1 - lollypop) on an emulator and low and behold, no buttons!

So we have an explanation, now to work on a solution ;)

**Comment by Matze2** [comment]
So it seems only these new CalculateButton components that give problems.

**Comment by S-Bartfast** [comment]
And yes, the "Plain" format does indeed work for me in Android 5.1 as well.

I really haven't tested the Plain format feature nearly as much as I have the 'button' formats so please do make note of any unusual behaviour you encounter and I'll see to it that I address anything I can.


All the best,
Slarti.

**Comment by Matze2** [comment]
Yes, I will prepare all my Multi caches that I plan to do next with parameterized coordinates and will give feedback. Need to watch your videos again tonight 👍 .
As it is possible to work in Plain mode, this not such a critical issue.

I think this is really a killer feature.

**Comment by S-Bartfast** [comment]
Still trying to work this one out I'm afraid.

I've managed to determined that it seems to be related to the "GridLayout" used to arrange the coordinate buttons, and I suspect it is the ` <item name="android:layout_columnWeight">1</item>` setting in the "button_digit" 'style' that is causing the issue but at this stage I'm unsure of how to fix it.  That line in particular does have a warning stating that API level 21 is required to use this feature and warns that the manifest states it supports down to API 14 however Android 5.1 is API 22, so that shouldn't be an issue.

Further more, I cant help but note that when creating the feature I added the line `compile 'com.android.support:gridlayout-v7:23.0.+'` to the 'build.gradle' file for the project which does look suspicious, and one suggestion for a similar problem I read online was to replace this with compile `'com.android.support:gridlayout-v7:22.0.+'`, which I was expecting to fix the problem (though be it only for API 22) however it did not seem to help, even for API 22.

Anyone with more experience familiar with this issue?

**Comment by Matze2** [comment]
Now that I used the calculator a few times in plain mode I am not really missing anything.

In the worst case you can just disable all modes but plain for API level <= 22.
This could be at least a temporary solution.

**Comment by S-Bartfast** [comment]
Yes, simply disabling the 'button formats' and forcing uses to enter in plain format would work as a stop gap, but Android 5.1 (lollipop) really doesn't seem that old??

**Comment by Lineflyer** [comment]
> In the worst case you can just disable all modes but plain for API level <= 22.

@S-Bartfast 
Did you test other API levels between 14-22
Is it only on API=22 or others as well?

I can try to look into my box of android testing devices and test on some API 14-22 to see where this problem exists.

**Comment by S-Bartfast** [comment]
22 was as far back as I went but I can't imagine anything earlier will work.  In fact I should imagine anything below 21 will additionally have issues with the "layout_columnWeight" as mentioned in the warning on top of whatever the Android 5.1 issue is.

**Comment by Lineflyer** [comment]
Ok, I will check if I own some real devices with API<22 to give it a test.
But then it seems we can only fallback to plain mode on API<22 indeed. :(

While plain mode works as it should, it is much less self explanatory which is a pitty for users on one of these versions.

**Comment by S-Bartfast** [comment]
Yeah, I'm sure there will be a solution to using grid layouts in older versions of Android, I just haven't figured it out yet.

**Comment by S-Bartfast** [comment]
By the way, API 14 is getting very old now.  We could potentially look at updating the minimum supported to something grater than 14 if it proves too hard to go back that far.

Here's the developer site showing a break down of the proportion of devices running different versions of Android:
https://developer.android.com/about/dashboards/index.html

And here's an image showing cumulative totals:
![image](https://user-images.githubusercontent.com/6573522/29778117-685911a6-8c4d-11e7-8523-420eceede0d4.png)

**Comment by Lineflyer** [comment]
Here are my test results from real devices using todays NB:
| Device  | Version/API | Result
| ------------- | ------------- |------------- |
| ZTE Blade S6 | 5.0.2 / 22  | No buttons, Plain mode OK |
| ZTE Grand Memo LTE | 4.1.2 / 16  | No buttons, Plain mode OK |
| HTC One Mini | 4.4.2 / 19  | No buttons, Plain mode OK |
| Samsung Galxy Ace | 6.0 / 23  | Works OK |

 So, indeed, it seems to be all versions below API 23 affected by this version.

Regarding the minimum supported version:
We raised it to API 14 onlyhalf a year ago (it was API 9 before). 

BTW:
I quickly checked our installation statistics (active installations as of 15.08.17):
- Installations API<14 (however this is possible) = 0,68%
- Installations 14<API<22 = 34,45%
- Installations API>22 = 64,87%

**Comment by Matze2** [comment]
Sorry, let me briefly interrupt your technical discussion (which I follow with big interest) to explain my personal opinion as an end user in more detail. I think my last statement was a little bit short and misunderstanding.

While the button issue might be an interesting technical challenge, the modification of the coordinate pattern is only one part of the big calculator feature. For me personally this is over-estimated. See the attached screenshot as a real example where the north coordinate consists of two variables and the east part consists of one.

I cannot imagine how you achieve this format with buttons in a more user-friendly then with just editing the coordinate directly. In my case I would always re-edit the auto-assigned variable names because I don*t want to mix them with the variables used in the listing. Usually in the listing they use abcdef so I use UVWXYZ for the coordinate pattern.

There are other useability gaps that I am missing more, e.g.
- the waypoint descriptions (sometimes the formula is there) is not copyable and therefore cannot be pasted to the calculator
- when you have a formula with only two or even one variable I had a xase where trailing zeroes are not printed in the result (cannot reproduce it at the moment)
- use of "capital" variables in formulas (it is needed to create temporary variables instead)
- and finally global variables to be shared over all waypoints (ok, this is really a big thing)

In other words: even the other modes will work I will probably stay with the "Plain" mode.

Sorry for the interruption,
Matthias
![screenshot_2017-08-28-19-54-40](https://user-images.githubusercontent.com/12542802/29788457-f3cb118c-8c32-11e7-9623-8d09962b1aaa.png)

**Comment by Lineflyer** [comment]
@Matze2 
Thanks for your feedback.
Yes, I agree, that the plain mode is quite usable. BTW: Also in button mode you can reassign the variable names.

However my main concern here is, that the calculator comes up unsuable on all devices API<22.
However I also think, that fixing it to plain mode on API<22 would be a quick workaround for now until we maybe find a better solution.

**Comment by Matze2** [comment]
Sorry for closing the issue accidentally. That was not my intention.

**Comment by S-Bartfast** [comment]
> We raised it to API 14 onlyhalf a year ago (it was API 9 before).

Wow, API 9!  How did you guys manage?
I guess API 14 must seem positively new compared to that ;)

And Matze2, I'm pleased to hear the 'Plain' mode is working well for you.  I really only added it for completeness as it was used in the regular coordinate input dialogue and wasn't expecting it to get much love.   As you say though, it seems to suit your needs quite well, possibly better than the regular buttons would have ;)

> * the waypoint description (sometimes the formula is there) is not copyable and therefore cannot be pasted to the calculator

Interesting.  I thought temporarily about linking the 'Calculator notes' with the 'waypoint notes' (or 'user notes') but didn't end up giving it much thought and just implemented a 3rd 'notes' field exclusively for the calculator.  Would you find it more useful to have these notes merged in some way?  If so I'll look at it as another ticket.

> * when you have a formula with only two or even one variable I had a case where trailing zeroes are not printed in the result (cannot reproduce it at the moment)

Yes, the calculator always adds 'leading-zeros' to pad-out values to the required number of digits.  I thought about adding trailing zeros to values that were to be placed after the decimal point (rather than leading zeros) which mathematically would be more correct but the one cache I did requiring 'post-decimal-padding' actually required _leading_ zeros anyhow, which seems to be how the regular 'coordinate-input' dialogue works as well, so I just left it.  Do you think trailing zero padding would be preferable after the decimal point?

> * use of "capital" variables in formulas (it is needed to create temporary variables instead)

Arr, yes.  The reason it's as it is is to alleviate issues with 'recursion'.  That is defining a value in relation to a variable, which in turn is defined in relation to the original value.  As in `A = B+1`, `B = 2*C` and `C = 1/A`.  There are ways to handle these things but they can get complicated as you tease out the details, which is why I opted for the simple architecture of 'Equations' (uppercase) only being able use 'Free Variables' (lowercase), which can only use arithmetic expressions.  Using this there is no way an expression can wind up referencing itself.

> * and finally global variables to be shared over all waypoints (ok, this is really a big thing)

It's not that big.  It will require some thought but is most definitely achievable.  I just have to work out how ;)

**Comment by Lineflyer** [comment]
> Would you find it more useful to have these notes merged in some way? If so I'll look at it as another ticket.

I agree, that should be a seperate ticket to discuss this in detail.

>  Do you think trailing zero padding would be preferable after the decimal point?

Trailing zeros are IMHO not desireable, if e.g. the cache owner has a formula like `N50 12.A` with `A=12` I would expect the solution is `N50 12.012' rather than 'N50 12.120`.
I know there are a few caches where cache owners expect the second solution to be the correct one, however this is mathematically not correct, isn't it?

> Arr, yes. The reason it's as it is is to alleviate issues with 'recursion'.

That is of course one important syntax rule you have to know when you are using the plain mode. In other modes, where you have the buttons, this is more easy to learn and understand.

> > and finally global variables to be shared over all waypoints (ok, this is really a big thing)

There is already #6691 for this.

**Comment by Matze2** [comment]
> the waypoint description (sometimes the formula is there) is not copyable and therefore cannot be pasted to the calculator

I will make up my mind tonight and create a separate ticket for that.

**Comment by Matze2** [comment]
I am sorry. I just saw that I can copy the waypoint note when I am inside the waypoint editor.
I tried that only in the waypoint overview and there I can only copy the coordinates.
Then this is no issue at all.

What I wonder if an extra note (it's the third one) in the calculator is really needed or if we cannot use the existing editable waypoint note for everything? Sometimes I feel that it is a disadvantage to not see the calculator note in the waypoint overview.

**Comment by SammysHP** [comment]
Regarding leading/trailing zeros: https://geoclub.de/forum/viewtopic.php?p=1272389#p1272389 (German)

**Comment by Matze2** [comment]
About the leading/trailing zeros: I faced a strange behavior once but cannot reproduce it at the moment.
So therefore no opinion on that at the moment.

**Comment by Lineflyer** [comment]
@S-Bartfast 
Regarding the original topic of this issue:
Do you think you can submit a PR for the workaround (plain only on API<22) or even a solution?

I would like to estimate a next release in September, where we shoud have a first version of the calculator without major issue.

BTW:
Will change the topic of this issue according to the current situation.

**Comment by Matze2** [comment]
I created also an issue for the trailing zeros interpretation now. Hope I don't spam you too much with all these calculator issues. :simple_smile:

**Comment by S-Bartfast** [comment]
> I created also an issue for the trailing zeros interpretation now.

Good work, thank you.
>Hope I don't spam you too much with all these calculator issues. :)

Not at all.  In fact your feedback is most appreciated!

Feel free to make more tickets for anything else you think could be improved as well.  We very much appreciate feedback, and your requests in particular have been nicely detailed and have actually been a pleasure to work with ;)

**Comment by Matze2** [comment]
I would love to also help with code contributions.
But I haven't had the time until now to introduce myself into the tool chain for app programming which is completely new to me.

**Comment by S-Bartfast** [comment]
Okay, finally sorted this one out.
It actually turned out to be quite simple to fix once I understood what the problem was.

I checked this fix against API 26 (Oreo), 23 (Marshmallow), 22 (Lollipop), 21 (Lollipop) and API 19 (KitKat), all of which work.

API 15 (Ice Cream Sandwich) has some issues in regard to 'ButterKnife' but it's an unrelated problem and will require its own ticket.

**Comment by Matze2** [comment]
Just a short notice. I see the buttons now, but have not experimented too much with it.
My concrete example has a formula that needs only two buttons instead of three at the end - similar to the screenshot above with UV and X. But I didn't manage to do the same with buttons.

The issue itself seems to be fixed now. Thank you very much!

**Comment by S-Bartfast** [comment]
> My concrete example has a formula that needs only two buttons instead of three at the end - similar to the screenshot above with UV and X. But I didn't manage to do the same with buttons.

Are you aware of the "long-press and enter SPACE" technique?  Or perhaps I've misunderstood what you mean.

In any event, feel free to report back with any findings after you've had a play :)

**Comment by Matze2** [comment]
Ok. Managed it. This is how it looks now.

![screenshot_2017-10-12-08-58-35](https://user-images.githubusercontent.com/12542802/31483237-cf19cc54-af2b-11e7-9ad5-9566c6656f1b.png)

Is this ok that the "long-pressed" buttons are grey  It looks a little bit strange that the X is grey without button frame (because I renamed it) and the Y is white with frame?

**Comment by pstorch** [comment]
> Are you aware of the "long-press and enter SPACE" technique?

I'm thinking about if we should change the button modes. With a short press it toggles between variable and original value. Maybe it should also toggle through the "space" case. So that this feature is more accessible.

**Comment by S-Bartfast** [comment]
> It looks a little bit strange that the X is grey without button frame (because I renamed it) and the Y is white with frame?

Yeah, the reason they use different colours is because the 'white' buttons are dependent on the previous letter (that is they are the letter that follows the last alphabetically).  So if you change the layout all the white letters can potentially change as well, where as the 'grey' buttons have their value fixed and will not change.

Originally I was trying to find a way to use an 'embossed' font to represent the set values but couldn't find a way to make it work so ended up settling for the 'steel' colour instead.

**Comment by Matze2** [comment]
While trying I found another issue: if I add a new waypoint and open the calculator first time it does not take over the coordinates from the waypoint. They differ in the last three digits.
If you want I can create a new issue tonight for this.

**Comment by S-Bartfast** [comment]
> if I add a new waypoint and open the calculator first time it does not take over the coordinates from the waypoint. They differ in the last three digits.

It should take on the values entered in the "Input Coordinates" dialogue, which can reflect the caches coordinates, your current position, or anything else you care to enter.

> If you want I can create a new issue tonight for this.

If you are getting values that differ from the "Input Coordinates" dialogue that is most definitely a bug and I will address it.

**Comment by Matze2** [comment]
I used GC6DW57, create a new waypoint, pressed "Cache coordinates".
They display shows 48 22.922 and 011 32.208.
Now I enter the calculator and the coordinates shows there are 48 22.055 and 011 32.012.

I am at work now and have no time to create a proper issue. But I can do tonight.

**Comment by Matze2** [comment]
Now my last post back on the topic:
> If you find the white letter offends you (as it does me) you can always long-press the 'Y' and manually specify its value as well thus making it grey too.

As you probably already noticed from my screenshots in almost every case I change the letters to the end of the alphabet, usually from U to Z. I do this because the listing formulas normally use ABC and so on, too, and I don't want to mix the letters for the digits with the variables for the formulas.
This means in my case the first letter is always grey (usually U or X) and the others are white.
Maybe the way how I use the calculator is special.

**Comment by S-Bartfast** [comment]
> Maybe the way how I use the calculator is special.

That is a sensible way to use the calculator.

I agree the 'namespace-clash' is unfortunate but I couldn't find a better way to deal with it so just hoped the upper and lower case designations would be enough.  I did actually consider having the 'white letters' start at "Z" and successively decrement their way down toward "A" (which would be somewhat more similar to your approach) but abandoned the idea.

While the current approach is clear (which is probably the most important thing) I agree it is somewhat unsightly and would prefer a cleaner solution.  Potentially I could set the last letter to be "Z" and 'back-track' toward the beginning reducing the letter each time.  That would produce exactly what you are after but I fear it would not be so intuitive to others, especially when custom letters need to be manually specified.

If the simple [A][B][C].[D][E] format won't suffice for your waypoint I'm afraid you'll have to either "manually specify" every letter (to make the all 'grey'), revert to the 'Plain Format' mode, or simply endure the inconsistent colours.  Sorry.

I would be more than happy to entertain other approaches though if anyone has suggestions.

**Comment by Matze2** [comment]
I agree going backward from Z is not intuitive. Currently I also have no better suggestion. But I am more than happy with the calculator feature.
I cannot count how often I had to visit a multi cache twice because I made a mistake on calculation. With the calculator preparation in advance this hasn't happened again until now. 👍

**Comment by S-Bartfast** [comment]
> Maybe it should also toggle through the "space" case. So that this feature is more accessible.

That definitely has some merit.
Personally I like the "simple toggle" case myself but agree the current "hide" functionality is hard to happen upon naturally.  Perhaps it would be worth sacrificing the 'simple toggle' cleanness in order to expose the 'hide' capability more.  I'll look into it.

**Comment by Lineflyer** [comment]
> Personally I like the "simple toggle" case myself but agree the current "hide" functionality is hard to happen upon naturally. Perhaps it would be worth sacrificing the 'simple toggle' cleanness in order to expose the 'hide' capability more. I'll look into it.

I agree and I like the proposal from @pstorch with toggeling through the "space case".
